internal/data/metric.go

```diff
@@ -0,0 +1,629 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package data
+
+import (
+	otlpmetric "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+)
+
+// This file defines in-memory data structures to represent metrics.
+// For the proto representation see https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/metrics/v1/metrics.proto
+
+// MetricData is the top-level struct that is propagated through the metrics pipeline.
+// This is the newer version of consumerdata.MetricsData, but uses more efficient
+// in-memory representation.
+type MetricData struct {
+	resourceMetrics []*ResourceMetrics
+}
+
+func NewMetricData(resourceMetrics []*ResourceMetrics) MetricData {
+	return MetricData{resourceMetrics}
+}
+
+// MetricCount calculates the total number of metrics.
+func (td MetricData) MetricCount() int {
+	metricCount := 0
+	for _, rml := range td.resourceMetrics {
+		metricCount += len(rml.metrics)
+	}
+	return metricCount
+}
+
+// A collection of metrics from a Resource.
+//
+// Must use NewResourceMetrics functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type ResourceMetrics struct {
+	// The resource for the metrics in this message.
+	// If this field is not set then no resource info is known.
+	resource *Resource
+
+	// A list of Spans that originate from a resource.
+	metrics []*Metric
+}
+
+func NewResourceMetrics(resource *Resource, metrics []*Metric) *ResourceMetrics {
+	return &ResourceMetrics{resource, metrics}
+}
+
+func (mr *ResourceMetrics) Resource() *Resource {
+	return mr.resource
+}
+
+func (mr *ResourceMetrics) SetResource(r *Resource) {
+	mr.resource = r
+}
+
+func (mr *ResourceMetrics) Metrics() []*Metric {
+	return mr.metrics
+}
+
+func (mr *ResourceMetrics) SetMetrics(s []*Metric) {
+	mr.metrics = s
+}
+
+// Metric defines a metric which has a descriptor and one or more timeseries points.
+//
+// Must use NewSMetric* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type Metric struct {
+	// Wrap OTLP Metric.
+	orig *otlpmetric.Metric
```

Comment:
It appears we override all fields below. I think `orig` is not needed at all.

Reply:
If we construct it from an already existing `orig *otlpmetric.Metric` for example received from otlp receiver, this can be used if we reconstruct the proto. Do you think that is not the behavior you want?

Reply:
OK, let's keep for now and see how it will be used during proto construction. Can be removed later if not needed.